File: lib/python/openlinktoken/src/main/openlinktoken/attributes/attribute_field.py
# SPDX-License-Identifier: MIT

from typing import TYPE_CHECKING, Type

if TYPE_CHECKING:
    from openlinktoken.attributes.attribute import Attribute


class AttributeField:
    """
    Represents a named field slot in a person record.

    An AttributeField separates two concerns that were previously conflated:
    - Field identity: "which value from the person record?" (the field_id)
    - Attribute behavior: "how to normalize/validate?" (the attribute_class)

    This allows multiple fields to share the same attribute type (e.g., two fields
    both using StringAttribute normalization) while remaining distinct keys in the
    person attributes map.
    """

    def __init__(self, field_id: str, attribute_class: Type["Attribute"]):
        """
        Create an attribute field with the given identity and attribute type.

        Args:
            field_id: Unique field identifier (e.g., "LastName", "MotherLastName").
            attribute_class: The attribute class providing normalization and validation behavior.
        """
        if field_id is None:
            raise ValueError("field_id must not be None")
        if attribute_class is None:
            raise ValueError("attribute_class must not be None")
        self._field_id = field_id
        self._attribute_class = attribute_class

    @property
    def field_id(self) -> str:
        """Get the unique field identifier."""
        return self._field_id

    @property
    def attribute_class(self) -> Type["Attribute"]:
        """Get the attribute class providing normalization and validation behavior."""
        return self._attribute_class

    def __eq__(self, other: object) -> bool:
        if self is other:
            return True
        if not isinstance(other, AttributeField):
            return False
        return self._field_id == other._field_id

    def __hash__(self) -> int:
        return hash(self._field_id)

    def __repr__(self) -> str:
        return f"AttributeField(field_id='{self._field_id}', attribute_class={self._attribute_class.__name__})"

File: lib/java/openlinktoken/src/main/java/org/openlinktoken/attributes/AttributeField.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.attributes;

import java.io.Serializable;
import java.util.Objects;

import lombok.Getter;

/**
 * Represents a named field slot in a person record.
 *
 * <p>
 * An {@code AttributeField} separates two concerns that were previously conflated:
 * <ul>
 *   <li><b>Field identity</b> — "which value from the person record?" (the {@code fieldId})</li>
 *   <li><b>Attribute behavior</b> — "how to normalize/validate?" (the {@code attributeClass})</li>
 * </ul>
 *
 * <p>
 * This allows multiple fields to share the same attribute type (e.g., two fields both using
 * {@code StringAttribute} normalization) while remaining distinct keys in the person attributes map.
 */
@Getter
public final class AttributeField implements Serializable {

    private static final long serialVersionUID = 1L;

    private final String fieldId;
    private final Class<? extends Attribute> attributeClass;

    /**
     * Creates an attribute field with the given identity and attribute type.
     *
     * @param fieldId        unique field identifier (e.g., "LastName", "MotherLastName")
     * @param attributeClass the attribute class providing normalization and validation behavior
     */
    public AttributeField(String fieldId, Class<? extends Attribute> attributeClass) {
        this.fieldId = Objects.requireNonNull(fieldId, "fieldId must not be null");
        this.attributeClass = Objects.requireNonNull(attributeClass, "attributeClass must not be null");
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof AttributeField other)) {
            return false;
        }
        return fieldId.equals(other.fieldId);
    }

    @Override
    public int hashCode() {
        return fieldId.hashCode();
    }

    @Override
    public String toString() {
        return "AttributeField{fieldId='" + fieldId + "', attributeClass=" + attributeClass.getSimpleName() + "}";
    }
}
